### ISSUE TITLE
Reflection incorrectly hides events

### ISSUE THREAD

**Issue by MaximLipnin** [mono/mono#11235]
## Steps to Reproduce
```c#
using System;
using System.Reflection;

namespace test
{
    public class MainClass
    {
        public static void Main(string[] args)
        {
            Type derived = typeof(Derived);
            EventInfo[] events = derived.GetEvents(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static);
            Console.WriteLine (events.Length);

            Type staticDerived2 = typeof(StaticDerived2);
            EventInfo[] events2 = staticDerived2.GetEvents(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static);
            Console.WriteLine (events2.Length);
        }

        private abstract class Base
        {
            public event Action MyEvent { add { } remove { } }
            public int MyProp { get; }
        }

        private abstract class Derived : Base
        {
            private new event Action MyEvent { add { } remove { } }
            private new int MyProp { get; }
        }

        private abstract class Derived2 : Derived
        {
        } 

        private class StaticBase
        {
            public event Action MyEvent { add { } remove { } }
        }

        private class StaticDerived : StaticBase
        {
            public new static event Action MyEvent { add { } remove { } }
        }

        private class StaticDerived2 : StaticDerived
        {
        }               
    }
}
```
or use [HideDetectionHappensBeforeBindingFlagChecks ](https://github.com/mono/corefx/blob/master/src/System.Runtime/tests/System/Reflection/TypeTests.Get.CornerCases.cs#L275) and [HideDetectionHappensBeforeStaticInNonFlattenedHierarchyChecks](https://github.com/mono/corefx/blob/master/src/System.Runtime/tests/System/Reflection/TypeTests.Get.CornerCases.cs#L303) CoreFX tests.

### Current Behavior

The actual output is
```
1
1
```

### Expected Behavior

The expected output is
```
0
0
```

## On which platforms did you notice this

[X] macOS
[ ] Linux
[X] Windows

**Version Used**:

5.16.0 / the latest master

## Stacktrace

```
Please paste the stack trace here if available.
```

**Comment by henricm** [comment]
I've debugged this in mono and .net core and there are some differences in how the event info is extracted from the type.

### Mono
In Mono this method is responsible `GetEvents_native`: 
https://github.com/mono/mono/blob/master/mono/metadata/icall.c#L4350

In the first test case, with the `Derived` type, which should hide the event `MyEvent`, `GetEvents_native` will return the event from the base class and it has no handling to check if the event is private in the start class. Another issue is that the flags seems to be set to `BindingFlags.Default` (shouldn't they be set to something?).

`GetEvents_native` is called from `rttype.cs` which is the same as .net core:

https://github.com/mono/mono/blob/master/mcs/class/referencesource/mscorlib/system/rttype.cs#L2992-L3017

For mono the event returned will be added to the `candidates` list since the `BindingFlags` is 0 (`Default`).

### .NET Core

For .net core this line returns one event info object:
```
            RuntimeEventInfo[] cache = Cache.GetEventList(listType, name);
```
However this seem to be the private event from the `Derived` class since it has `BindingFlags` set to  `Instance | NonPublic`:

![image](https://user-images.githubusercontent.com/3715975/48552388-bd38db00-e8d8-11e8-888f-e8ba1e539928.png)

The code will then filter this out in the following loop where the flags are checked to match the flags sent to the function
.
### Summary

To sum up, the native code in mono will fetch the public event from the base class instead of the private one in the start class, where .net returns the private event in the start class. 
Should we change the behaviour of the native code in `GetEvents_native` and how?

Or is the long term solution to try to use more of the managed code in rttype.cs?

**Comment by marek-safar** [comment]
I think at this point we should change the unmanaged implementation logic to confirm the expected behaviour

**Comment by henricm** [comment]
Should we change the behaviour of the unmanaged code so we do not do any filtering based on the `bindingAttr` flags at all? To make it more similar to the .net code?
In that case we could remove the `bindingAttr` parameter to `GetEvents_native` entirely, but I'm not sure I have the whole picture to see the implications for such change.

**Comment by lambdageek** [comment]
The .NET Core code for `Cache.GetEventList(listType, name);` calls down to 
https://github.com/dotnet/coreclr/blob/6d9fc60df3b721eab98ca08386a1f01358f93839/src/System.Private.CoreLib/src/System/RtType.cs#L1150

which does do some filtering, but doesn't apply the binding flags.

We should probably mimic this behavior.

**Comment by henricm** [comment]
Thanks. We also need implementation of BindingFlags property in MonoEvent.RuntimeEventInfo:
https://github.com/mono/mono/blob/master/mcs/class/corlib/System.Reflection/MonoEvent.cs#L68
since it's crucial for the filtering based on `bindAttr` in the managed code.
Will try to get that in place.
